internal/storage/download.go

```diff
@@ -101,51 +98,108 @@ func downloadArtifact(to string, artifact *Artifact, progress progressBytes, ser
 	if err != nil {
-		return err
+		return 0, err
 	}
 	defer response.Body.Close()
 
-	// HTTP Status code is NOT in the 2xx range
-	if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusMultipleChoices {
-		return fmt.Errorf("http status code is not in the 2xx range: %v", response.StatusCode)
-	}
-
 	// Check if HTTP server support Range header. If not, delete existing file and perform regular download
 	if response.Header.Get("Accept-Ranges") != "bytes" || response.Header.Get("Content-Range") == "" {
-		logger.Infof("Resume is not supported, remove previous file: %s", to)
+		logger.Infof("resume is not supported, remove previous file: %s", to)
 		if err := os.Remove(to); err != nil {
-			return err
+			logger.Errorf("error removing partially downloaded file %s", to)
+			return 0, err
 		}
-		return download(to, response.Body, artifact, progress, done)
+		return download(to, response.Body, artifact, progress, serverCert, retryCount, retryInterval, done)
 	}
 
 	// Download the rest of the file.
-	logger.Debugf("Resume previous download of file: %s", to)
+	logger.Debugf("resume previous download of file: %s", to)
 	file, err := os.OpenFile(to, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0755)
 	if err != nil {
-		return err
+		return 0, err
 	}
 	defer file.Close()
 
 	if progress != nil {
 		progress(offset)
 	}
-	if w, err := copy(file, response.Body, int64(artifact.Size)-offset, progress, done); err != nil {
-		logger.Debugf("Written bytes: %v", w)
-		return err
+	return downloadFile(file, response.Body, to, offset, artifact, progress, serverCert, retryCount, retryInterval, done)
+}
+
+func downloadFile(file *os.File, input io.ReadCloser, to string, offset int64, artifact *Artifact,
+	progress progressBytes, serverCert string, retryCount int, retryInterval int, done chan struct{}) (int64, error) {
+	w, err := copy(file, input, int64(artifact.Size)-offset, progress, done)
+	if err == nil {
+		err = validate(to, artifact.HashType, artifact.HashValue)
+		offset = 0 // in case of error, re-download the file
+		w = 0
+	} else {
+		logger.Debugf("written bytes: %v", w)
+		offset += w
+	}
+	if err == nil {
+		return w, nil
+	}
+	retryCount--
+	for retryCount >= 0 {
+		var deltaBytes int64
+		logger.Errorf("error copying artifact %s, remaining attempts - %d, cause: %v", file.Name(), retryCount, err)
+		logger.Infof("%d seconds timeout until next attempt", retryInterval)
+		file.Close()
+		time.Sleep(time.Duration(retryInterval) * time.Second)
+		logger.Infof("retrying to download artifact %s, current bytes written - %d", file.Name(), offset)
+		deltaBytes, err = resume(to, offset, artifact, progress, serverCert, 0, 0, done)
+		if err == nil {
+			break
+		}
+		offset += deltaBytes
+		retryCount--
+	}
+	return w, err
+}
+
+func getDownloadResponse(link string, offset int64, serverCert string, retryCount *int, retryInterval int) (*http.Response, error) {
```

Comment:
retryCount - to use method argument as a output parameter is not best possible practice. In case of an error, the retry can be decreased in the caller.